File: sw/cpp/src/protocol.cpp
#include "reference/protocol.hpp"

#include <algorithm>
#include <array>
#include <limits>

#include "reference/crc32c.hpp"

namespace reference {
namespace {

constexpr std::array<std::uint8_t, 2> sof{0xA5U, 0x5AU};
constexpr std::size_t header_bytes = 6U;
constexpr std::size_t crc_bytes = 4U;
constexpr std::size_t minimum_frame_bytes = sof.size() + header_bytes + crc_bytes;

void append_u16(Bytes& output, const std::uint16_t value) {
    output.push_back(static_cast<std::uint8_t>(value & 0xFFU));
    output.push_back(static_cast<std::uint8_t>((value >> 8U) & 0xFFU));
}

void append_u32(Bytes& output, const std::uint32_t value) {
    for (unsigned shift = 0; shift < 32U; shift += 8U) {
        output.push_back(static_cast<std::uint8_t>((value >> shift) & 0xFFU));
    }
}

std::uint16_t read_u16(const std::span<const std::uint8_t> data, const std::size_t offset) {
    return static_cast<std::uint16_t>(
        data[offset] | (static_cast<std::uint16_t>(data[offset + 1U]) << 8U)
    );
}

std::uint32_t read_u32(const std::span<const std::uint8_t> data, const std::size_t offset) {
    std::uint32_t value = 0;
    for (unsigned index = 0; index < 4U; ++index) {
        value |= static_cast<std::uint32_t>(data[offset + index]) << (index * 8U);
    }
    return value;
}

}

Bytes serialize_frame(const Frame& frame) {
    if (frame.payload.size() > max_payload_bytes) {
        throw FrameLengthError("payload exceeds Frame V0 maximum");
    }
    if (frame.payload.size() > std::numeric_limits<std::uint16_t>::max()) {
        throw FrameLengthError("payload cannot be represented by length field");
    }
    Bytes output{sof.begin(), sof.end()};
    output.push_back(frame.version);
    output.push_back(frame.message_type);
    append_u16(output, frame.flags);
    append_u16(output, static_cast<std::uint16_t>(frame.payload.size()));
    output.insert(output.end(), frame.payload.begin(), frame.payload.end());
    const auto crc_input = std::span<const std::uint8_t>(output).subspan(sof.size());
    append_u32(output, crc32c(crc_input));
    return output;
}

Frame parse_frame(const std::span<const std::uint8_t> data) {
    if (data.size() < minimum_frame_bytes) {
        throw FrameLengthError("frame is shorter than minimum size");
    }
    if (!std::equal(sof.begin(), sof.end(), data.begin())) {
        throw ProtocolError("frame does not start with SOF");
    }
    const auto version = data[2];
    const auto message_type = data[3];
    const auto flags = read_u16(data, 4);
    const auto payload_length = read_u16(data, 6);
    if (payload_length > max_payload_bytes) {
        throw FrameLengthError("payload length exceeds Frame V0 maximum");
    }
    const auto expected_size = minimum_frame_bytes + payload_length;
    if (data.size() != expected_size) {
        throw FrameLengthError("frame length does not match payload length");
    }
    if (version != frame_version) {
        throw FrameVersionError("unsupported Frame version");
    }
    const auto crc_offset = sof.size() + header_bytes + payload_length;
    const auto crc_input = data.subspan(sof.size(), header_bytes + payload_length);
    if (crc32c(crc_input) != read_u32(data, crc_offset)) {
        throw FrameCrcError("CRC mismatch");
    }
    const auto payload_begin = data.begin() + static_cast<std::ptrdiff_t>(sof.size() + header_bytes);
    const auto payload_end = payload_begin + payload_length;
    return {message_type, Bytes(payload_begin, payload_end), flags, version};
}

std::vector<Frame> FrameParser::feed(const std::span<const std::uint8_t> data) {
    buffer_.insert(buffer_.end(), data.begin(), data.end());
    std::vector<Frame> frames;
    while (true) {
        const auto sof_position = std::search(buffer_.begin(), buffer_.end(), sof.begin(), sof.end());
        if (sof_position == buffer_.end()) {
            const auto keep_first_sof_byte = !buffer_.empty() && buffer_.back() == sof.front();
            buffer_ = keep_first_sof_byte ? Bytes{sof.front()} : Bytes{};
            break;
        }
        buffer_.erase(buffer_.begin(), sof_position);
        if (buffer_.size() < sof.size() + header_bytes) {
            break;
        }
        const auto payload_length = read_u16(buffer_, 6);
        if (payload_length > max_payload_bytes) {
            errors_.emplace_back("payload length exceeds maximum");
            buffer_.erase(buffer_.begin());
            continue;
        }
        const auto frame_length = minimum_frame_bytes + payload_length;
        if (buffer_.size() < frame_length) {
            break;
        }
        try {
            frames.push_back(parse_frame(std::span<const std::uint8_t>(buffer_).first(frame_length)));
            buffer_.erase(buffer_.begin(), buffer_.begin() + static_cast<std::ptrdiff_t>(frame_length));
        } catch (const ProtocolError& error) {
            errors_.push_back(error.what());
            buffer_.erase(buffer_.begin());
        }
    }
    return frames;
}

std::size_t FrameParser::buffered_bytes() const noexcept {
    return buffer_.size();
}

const std::vector<std::string>& FrameParser::errors() const noexcept {
    return errors_;
}

}

File: sw/common/protocol.py
"""Test-only Frame V0 serializer and incremental parser."""

import struct
from dataclasses import dataclass
from enum import IntEnum

from . import protocol_constants as constants
from .crc32c import crc32c

SOF = constants.FRAME_SOF
FRAME_VERSION = constants.FRAME_VERSION
MAX_PAYLOAD_BYTES = constants.FRAME_MAX_PAYLOAD_BYTES
_HEADER = struct.Struct("<BBHH")
_CRC = struct.Struct("<I")
MIN_FRAME_BYTES = len(SOF) + _HEADER.size + _CRC.size


class ProtocolError(ValueError):
    """Base class for invalid Frame V0 input."""


class FrameLengthError(ProtocolError):
    """Report a Frame V0 length violation."""


class FrameCrcError(ProtocolError):
    """Report a Frame V0 CRC mismatch."""


class FrameVersionError(ProtocolError):
    """Report an unsupported Frame V0 version."""


class MessageType(IntEnum):
    """Identify test-only Frame V0 message types."""

    PING = constants.MESSAGE_TYPE_PING
    PONG = constants.MESSAGE_TYPE_PONG
    TOKEN_REQUEST = constants.MESSAGE_TYPE_TOKEN_REQUEST
    TOKEN_RESULT = constants.MESSAGE_TYPE_TOKEN_RESULT
    # Frame V0 used TEST names before the Token payload contract was fixed.
    TEST_REQUEST = TOKEN_REQUEST
    TEST_RESULT = TOKEN_RESULT
    ERROR_RESPONSE = constants.MESSAGE_TYPE_ERROR_RESPONSE


@dataclass(frozen=True, slots=True)
class Frame:
    """Represent a validated test-only Frame V0 payload and header.

    Attributes
    ----------
    message_type : int
        Unsigned 8-bit test message type.
    payload : bytes
        Opaque payload of at most 1024 bytes.
    flags : int
        Unsigned 16-bit flags field.
    version : int
        Unsigned 8-bit protocol version.
    """

    message_type: int
    payload: bytes = b""
    flags: int = 0
    version: int = FRAME_VERSION

    def __post_init__(self) -> None:
        if not 0 <= self.version <= 0xFF:
            raise ValueError("version must be an unsigned 8-bit value")
        if not 0 <= self.message_type <= 0xFF:
            raise ValueError("message_type must be an unsigned 8-bit value")
        if not 0 <= self.flags <= 0xFFFF:
            raise ValueError("flags must be an unsigned 16-bit value")
        if len(self.payload) > MAX_PAYLOAD_BYTES:
            raise FrameLengthError("payload exceeds Frame V0 maximum")


def serialize_frame(frame: Frame) -> bytes:
    """Serialize a Frame V0 value with little-endian fields and CRC-32C.

    Parameters
    ----------
    frame : Frame
        Validated frame fields and opaque payload.

    Returns
    -------
    bytes
        Complete wire representation including SOF and CRC.
    """
    header = _HEADER.pack(
        frame.version,
        frame.message_type,
        frame.flags,
        len(frame.payload),
    )
    crc_input = header + frame.payload
    return SOF + crc_input + _CRC.pack(crc32c(crc_input))


def parse_frame(data: bytes) -> Frame:
    """Parse and validate exactly one complete Frame V0 value.

    Parameters
    ----------
    data : bytes
        Complete serialized frame.

    Returns
    -------
    Frame
        Parsed header fields and opaque payload.

    Raises
    ------
    ProtocolError
        If framing, version, length, or CRC validation fails.
    """
    if len(data) < MIN_FRAME_BYTES:
        raise FrameLengthError("frame is shorter than the minimum size")
    if not data.startswith(SOF):
        raise ProtocolError("frame does not start with SOF")

    version, message_type, flags, payload_length = _HEADER.unpack_from(data, len(SOF))
    if payload_length > MAX_PAYLOAD_BYTES:
        raise FrameLengthError("payload length exceeds Frame V0 maximum")

    expected_length = MIN_FRAME_BYTES + payload_length
    if len(data) != expected_length:
        raise FrameLengthError("frame length does not match payload length")
    if version != FRAME_VERSION:
        raise FrameVersionError(f"unsupported Frame version: {version}")

    crc_input_end = len(SOF) + _HEADER.size + payload_length
    crc_input = data[len(SOF) : crc_input_end]
    expected_crc = _CRC.unpack_from(data, crc_input_end)[0]
    actual_crc = crc32c(crc_input)
    if actual_crc != expected_crc:
        raise FrameCrcError(
            f"CRC mismatch: expected 0x{expected_crc:08x}, got 0x{actual_crc:08x}"
        )

    payload = data[len(SOF) + _HEADER.size : crc_input_end]
    return Frame(
        version=version,
        message_type=message_type,
        flags=flags,
        payload=payload,
    )


class FrameParser:
    """Incrementally extract valid frames and resynchronize after bad input.

    Attributes
    ----------
    errors : list[ProtocolError]
        Validation errors observed while searching for the next valid SOF.
    """

    def __init__(self) -> None:
        self._buffer = bytearray()
        self.errors: list[ProtocolError] = []

    @property
    def buffered_bytes(self) -> int:
        """Return the number of bytes waiting for a complete frame.

        Returns
        -------
        int
            Number of bytes retained by the incremental parser.
        """
        return len(self._buffer)

    def feed(self, data: bytes) -> list[Frame]:
        """Consume stream bytes and extract every newly complete valid frame.

        Parameters
        ----------
        data : bytes
            Next chunk from the byte stream.

        Returns
        -------
        list[Frame]
            Valid frames completed by this input chunk.
        """
        self._buffer.extend(data)
        frames: list[Frame] = []

        while True:
            sof_index = self._buffer.find(SOF)
            if sof_index < 0:
                self._buffer[:] = self._buffer[-1:] if self._buffer.endswith(SOF[:1]) else b""
                break
            if sof_index:
                del self._buffer[:sof_index]
            if len(self._buffer) < len(SOF) + _HEADER.size:
                break

            version, _, _, payload_length = _HEADER.unpack_from(self._buffer, len(SOF))
            if payload_length > MAX_PAYLOAD_BYTES:
                self.errors.append(FrameLengthError("payload length exceeds maximum"))
                del self._buffer[0]
                continue

            frame_length = MIN_FRAME_BYTES + payload_length
            if len(self._buffer) < frame_length:
                break

            candidate = bytes(self._buffer[:frame_length])
            try:
                frame = parse_frame(candidate)
            except ProtocolError as error:
                self.errors.append(error)
                del self._buffer[0]
                continue

            if version != FRAME_VERSION:
                raise AssertionError("parse_frame accepted an unsupported version")
            frames.append(frame)
            del self._buffer[:frame_length]

        return frames
